{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d7e4358",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70b97579",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "c5b2042e",
   "metadata": {},
   "source": [
    "```bash\n",
    "pip install sqlalchemy psycopg2-binary \n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a1096f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f569baa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "engine = create_engine('postgresql://root:root@localhost:5432/ny_taxi')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "523f5cf5",
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "(psycopg2.OperationalError) connection to server at \"localhost\" (::1), port 5432 failed: FATAL:  password authentication failed for user \"root\"\n\n(Background on this error at: https://sqlalche.me/e/14/e3q8)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:3280\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[1;34m(self, fn, connection)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3278'>3279</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3279'>3280</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m fn()\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3280'>3281</a>\u001b[0m \u001b[39mexcept\u001b[39;00m dialect\u001b[39m.\u001b[39mdbapi\u001b[39m.\u001b[39mError \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:310\u001b[0m, in \u001b[0;36mPool.connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=302'>303</a>\u001b[0m \u001b[39m\"\"\"Return a DBAPI connection from the pool.\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=303'>304</a>\u001b[0m \n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=304'>305</a>\u001b[0m \u001b[39mThe connection is instrumented such that when its\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=307'>308</a>\u001b[0m \n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=308'>309</a>\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=309'>310</a>\u001b[0m \u001b[39mreturn\u001b[39;00m _ConnectionFairy\u001b[39m.\u001b[39;49m_checkout(\u001b[39mself\u001b[39;49m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:868\u001b[0m, in \u001b[0;36m_ConnectionFairy._checkout\u001b[1;34m(cls, pool, threadconns, fairy)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=866'>867</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m fairy:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=867'>868</a>\u001b[0m     fairy \u001b[39m=\u001b[39m _ConnectionRecord\u001b[39m.\u001b[39;49mcheckout(pool)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=869'>870</a>\u001b[0m     fairy\u001b[39m.\u001b[39m_pool \u001b[39m=\u001b[39m pool\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:476\u001b[0m, in \u001b[0;36m_ConnectionRecord.checkout\u001b[1;34m(cls, pool)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=473'>474</a>\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=474'>475</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcheckout\u001b[39m(\u001b[39mcls\u001b[39m, pool):\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=475'>476</a>\u001b[0m     rec \u001b[39m=\u001b[39m pool\u001b[39m.\u001b[39;49m_do_get()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=476'>477</a>\u001b[0m     \u001b[39mtry\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\impl.py:146\u001b[0m, in \u001b[0;36mQueuePool._do_get\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=144'>145</a>\u001b[0m         \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39msafe_reraise():\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=145'>146</a>\u001b[0m             \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dec_overflow()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=146'>147</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[1;34m(self, type_, value, traceback)\u001b[0m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=68'>69</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwarn_only:\n\u001b[1;32m---> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=69'>70</a>\u001b[0m         compat\u001b[39m.\u001b[39;49mraise_(\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=70'>71</a>\u001b[0m             exc_value,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=71'>72</a>\u001b[0m             with_traceback\u001b[39m=\u001b[39;49mexc_tb,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=72'>73</a>\u001b[0m         )\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=73'>74</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\compat.py:207\u001b[0m, in \u001b[0;36mraise_\u001b[1;34m(***failed resolving arguments***)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=205'>206</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=206'>207</a>\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=207'>208</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=208'>209</a>\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=209'>210</a>\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=210'>211</a>\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\impl.py:143\u001b[0m, in \u001b[0;36mQueuePool._do_get\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=141'>142</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=142'>143</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_create_connection()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=143'>144</a>\u001b[0m \u001b[39mexcept\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:256\u001b[0m, in \u001b[0;36mPool._create_connection\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=253'>254</a>\u001b[0m \u001b[39m\"\"\"Called by subclasses to create a new ConnectionRecord.\"\"\"\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=255'>256</a>\u001b[0m \u001b[39mreturn\u001b[39;00m _ConnectionRecord(\u001b[39mself\u001b[39;49m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:371\u001b[0m, in \u001b[0;36m_ConnectionRecord.__init__\u001b[1;34m(self, pool, connect)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=369'>370</a>\u001b[0m \u001b[39mif\u001b[39;00m connect:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=370'>371</a>\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__connect()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=371'>372</a>\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfinalize_callback \u001b[39m=\u001b[39m deque()\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:666\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=664'>665</a>\u001b[0m     \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39msafe_reraise():\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=665'>666</a>\u001b[0m         pool\u001b[39m.\u001b[39mlogger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mError on connect(): \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=666'>667</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=667'>668</a>\u001b[0m     \u001b[39m# in SQLAlchemy 1.4 the first_connect event is not used by\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=668'>669</a>\u001b[0m     \u001b[39m# the engine, so this will usually not be set\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[1;34m(self, type_, value, traceback)\u001b[0m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=68'>69</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwarn_only:\n\u001b[1;32m---> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=69'>70</a>\u001b[0m         compat\u001b[39m.\u001b[39;49mraise_(\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=70'>71</a>\u001b[0m             exc_value,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=71'>72</a>\u001b[0m             with_traceback\u001b[39m=\u001b[39;49mexc_tb,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=72'>73</a>\u001b[0m         )\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=73'>74</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\compat.py:207\u001b[0m, in \u001b[0;36mraise_\u001b[1;34m(***failed resolving arguments***)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=205'>206</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=206'>207</a>\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=207'>208</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=208'>209</a>\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=209'>210</a>\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=210'>211</a>\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:661\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=659'>660</a>\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstarttime \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime()\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=660'>661</a>\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdbapi_connection \u001b[39m=\u001b[39m connection \u001b[39m=\u001b[39m pool\u001b[39m.\u001b[39;49m_invoke_creator(\u001b[39mself\u001b[39;49m)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=661'>662</a>\u001b[0m pool\u001b[39m.\u001b[39mlogger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mCreated new connection \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, connection)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\create.py:590\u001b[0m, in \u001b[0;36mcreate_engine.<locals>.connect\u001b[1;34m(connection_record)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/create.py?line=588'>589</a>\u001b[0m             \u001b[39mreturn\u001b[39;00m connection\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/create.py?line=589'>590</a>\u001b[0m \u001b[39mreturn\u001b[39;00m dialect\u001b[39m.\u001b[39;49mconnect(\u001b[39m*\u001b[39;49mcargs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mcparams)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\default.py:597\u001b[0m, in \u001b[0;36mDefaultDialect.connect\u001b[1;34m(self, *cargs, **cparams)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=594'>595</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39mcargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mcparams):\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=595'>596</a>\u001b[0m     \u001b[39m# inherits the docstring from interfaces.Dialect.connect\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=596'>597</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdbapi\u001b[39m.\u001b[39;49mconnect(\u001b[39m*\u001b[39;49mcargs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mcparams)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\psycopg2\\__init__.py:122\u001b[0m, in \u001b[0;36mconnect\u001b[1;34m(dsn, connection_factory, cursor_factory, **kwargs)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=120'>121</a>\u001b[0m dsn \u001b[39m=\u001b[39m _ext\u001b[39m.\u001b[39mmake_dsn(dsn, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=121'>122</a>\u001b[0m conn \u001b[39m=\u001b[39m _connect(dsn, connection_factory\u001b[39m=\u001b[39;49mconnection_factory, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwasync)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=122'>123</a>\u001b[0m \u001b[39mif\u001b[39;00m cursor_factory \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[1;31mOperationalError\u001b[0m: connection to server at \"localhost\" (::1), port 5432 failed: FATAL:  password authentication failed for user \"root\"\n",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\HP\\Desktop\\data-engineering-zoomcamp\\week_1_basics_n_setup\\2_docker_sql\\pg-test-connection.ipynb Cell 6'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/HP/Desktop/data-engineering-zoomcamp/week_1_basics_n_setup/2_docker_sql/pg-test-connection.ipynb#ch0000004?line=0'>1</a>\u001b[0m engine\u001b[39m.\u001b[39;49mconnect()\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:3234\u001b[0m, in \u001b[0;36mEngine.connect\u001b[1;34m(self, close_with_result)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3218'>3219</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m, close_with_result\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m):\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3219'>3220</a>\u001b[0m     \u001b[39m\"\"\"Return a new :class:`_engine.Connection` object.\u001b[39;00m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3220'>3221</a>\u001b[0m \n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3221'>3222</a>\u001b[0m \u001b[39m    The :class:`_engine.Connection` object is a facade that uses a DBAPI\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3230'>3231</a>\u001b[0m \n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3231'>3232</a>\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3233'>3234</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_connection_cls(\u001b[39mself\u001b[39;49m, close_with_result\u001b[39m=\u001b[39;49mclose_with_result)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:96\u001b[0m, in \u001b[0;36mConnection.__init__\u001b[1;34m(self, engine, connection, close_with_result, _branch_from, _execution_options, _dispatch, _has_events, _allow_revalidate)\u001b[0m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=90'>91</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_has_events \u001b[39m=\u001b[39m _branch_from\u001b[39m.\u001b[39m_has_events\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=91'>92</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=92'>93</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dbapi_connection \u001b[39m=\u001b[39m (\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=93'>94</a>\u001b[0m         connection\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=94'>95</a>\u001b[0m         \u001b[39mif\u001b[39;00m connection \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m---> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=95'>96</a>\u001b[0m         \u001b[39melse\u001b[39;00m engine\u001b[39m.\u001b[39;49mraw_connection()\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=96'>97</a>\u001b[0m     )\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=98'>99</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_transaction \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_nested_transaction \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=99'>100</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__savepoint_seq \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:3313\u001b[0m, in \u001b[0;36mEngine.raw_connection\u001b[1;34m(self, _connection)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3290'>3291</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mraw_connection\u001b[39m(\u001b[39mself\u001b[39m, _connection\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3291'>3292</a>\u001b[0m     \u001b[39m\"\"\"Return a \"raw\" DBAPI connection from the connection pool.\u001b[39;00m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3292'>3293</a>\u001b[0m \n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3293'>3294</a>\u001b[0m \u001b[39m    The returned object is a proxied version of the DBAPI\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3310'>3311</a>\u001b[0m \n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3311'>3312</a>\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3312'>3313</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_wrap_pool_connect(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpool\u001b[39m.\u001b[39;49mconnect, _connection)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:3283\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[1;34m(self, fn, connection)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3280'>3281</a>\u001b[0m \u001b[39mexcept\u001b[39;00m dialect\u001b[39m.\u001b[39mdbapi\u001b[39m.\u001b[39mError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3281'>3282</a>\u001b[0m     \u001b[39mif\u001b[39;00m connection \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3282'>3283</a>\u001b[0m         Connection\u001b[39m.\u001b[39;49m_handle_dbapi_exception_noconnection(\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3283'>3284</a>\u001b[0m             e, dialect, \u001b[39mself\u001b[39;49m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3284'>3285</a>\u001b[0m         )\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3285'>3286</a>\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3286'>3287</a>\u001b[0m         util\u001b[39m.\u001b[39mraise_(\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3287'>3288</a>\u001b[0m             sys\u001b[39m.\u001b[39mexc_info()[\u001b[39m1\u001b[39m], with_traceback\u001b[39m=\u001b[39msys\u001b[39m.\u001b[39mexc_info()[\u001b[39m2\u001b[39m]\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3288'>3289</a>\u001b[0m         )\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:2117\u001b[0m, in \u001b[0;36mConnection._handle_dbapi_exception_noconnection\u001b[1;34m(cls, e, dialect, engine)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2114'>2115</a>\u001b[0m     util\u001b[39m.\u001b[39mraise_(newraise, with_traceback\u001b[39m=\u001b[39mexc_info[\u001b[39m2\u001b[39m], from_\u001b[39m=\u001b[39me)\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2115'>2116</a>\u001b[0m \u001b[39melif\u001b[39;00m should_wrap:\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2116'>2117</a>\u001b[0m     util\u001b[39m.\u001b[39;49mraise_(\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2117'>2118</a>\u001b[0m         sqlalchemy_exception, with_traceback\u001b[39m=\u001b[39;49mexc_info[\u001b[39m2\u001b[39;49m], from_\u001b[39m=\u001b[39;49me\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2118'>2119</a>\u001b[0m     )\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2119'>2120</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=2120'>2121</a>\u001b[0m     util\u001b[39m.\u001b[39mraise_(exc_info[\u001b[39m1\u001b[39m], with_traceback\u001b[39m=\u001b[39mexc_info[\u001b[39m2\u001b[39m])\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\compat.py:207\u001b[0m, in \u001b[0;36mraise_\u001b[1;34m(***failed resolving arguments***)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=203'>204</a>\u001b[0m     exception\u001b[39m.\u001b[39m__cause__ \u001b[39m=\u001b[39m replace_context\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=205'>206</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=206'>207</a>\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=207'>208</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=208'>209</a>\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=209'>210</a>\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=210'>211</a>\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=211'>212</a>\u001b[0m     \u001b[39mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\base.py:3280\u001b[0m, in \u001b[0;36mEngine._wrap_pool_connect\u001b[1;34m(self, fn, connection)\u001b[0m\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3277'>3278</a>\u001b[0m dialect \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdialect\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3278'>3279</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3279'>3280</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m fn()\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3280'>3281</a>\u001b[0m \u001b[39mexcept\u001b[39;00m dialect\u001b[39m.\u001b[39mdbapi\u001b[39m.\u001b[39mError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m   <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/base.py?line=3281'>3282</a>\u001b[0m     \u001b[39mif\u001b[39;00m connection \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:310\u001b[0m, in \u001b[0;36mPool.connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=301'>302</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=302'>303</a>\u001b[0m     \u001b[39m\"\"\"Return a DBAPI connection from the pool.\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=303'>304</a>\u001b[0m \n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=304'>305</a>\u001b[0m \u001b[39m    The connection is instrumented such that when its\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=307'>308</a>\u001b[0m \n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=308'>309</a>\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=309'>310</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m _ConnectionFairy\u001b[39m.\u001b[39;49m_checkout(\u001b[39mself\u001b[39;49m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:868\u001b[0m, in \u001b[0;36m_ConnectionFairy._checkout\u001b[1;34m(cls, pool, threadconns, fairy)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=864'>865</a>\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=865'>866</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_checkout\u001b[39m(\u001b[39mcls\u001b[39m, pool, threadconns\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, fairy\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=866'>867</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m fairy:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=867'>868</a>\u001b[0m         fairy \u001b[39m=\u001b[39m _ConnectionRecord\u001b[39m.\u001b[39;49mcheckout(pool)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=869'>870</a>\u001b[0m         fairy\u001b[39m.\u001b[39m_pool \u001b[39m=\u001b[39m pool\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=870'>871</a>\u001b[0m         fairy\u001b[39m.\u001b[39m_counter \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:476\u001b[0m, in \u001b[0;36m_ConnectionRecord.checkout\u001b[1;34m(cls, pool)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=473'>474</a>\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=474'>475</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mcheckout\u001b[39m(\u001b[39mcls\u001b[39m, pool):\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=475'>476</a>\u001b[0m     rec \u001b[39m=\u001b[39m pool\u001b[39m.\u001b[39;49m_do_get()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=476'>477</a>\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=477'>478</a>\u001b[0m         dbapi_connection \u001b[39m=\u001b[39m rec\u001b[39m.\u001b[39mget_connection()\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\impl.py:146\u001b[0m, in \u001b[0;36mQueuePool._do_get\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=143'>144</a>\u001b[0m     \u001b[39mexcept\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=144'>145</a>\u001b[0m         \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39msafe_reraise():\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=145'>146</a>\u001b[0m             \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_dec_overflow()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=146'>147</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=147'>148</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_do_get()\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[1;34m(self, type_, value, traceback)\u001b[0m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=67'>68</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m  \u001b[39m# remove potential circular references\u001b[39;00m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=68'>69</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwarn_only:\n\u001b[1;32m---> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=69'>70</a>\u001b[0m         compat\u001b[39m.\u001b[39;49mraise_(\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=70'>71</a>\u001b[0m             exc_value,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=71'>72</a>\u001b[0m             with_traceback\u001b[39m=\u001b[39;49mexc_tb,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=72'>73</a>\u001b[0m         )\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=73'>74</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=74'>75</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m compat\u001b[39m.\u001b[39mpy3k \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info[\u001b[39m1\u001b[39m]:\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=75'>76</a>\u001b[0m         \u001b[39m# emulate Py3K's behavior of telling us when an exception\u001b[39;00m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=76'>77</a>\u001b[0m         \u001b[39m# occurs in an exception handler.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\compat.py:207\u001b[0m, in \u001b[0;36mraise_\u001b[1;34m(***failed resolving arguments***)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=203'>204</a>\u001b[0m     exception\u001b[39m.\u001b[39m__cause__ \u001b[39m=\u001b[39m replace_context\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=205'>206</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=206'>207</a>\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=207'>208</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=208'>209</a>\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=209'>210</a>\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=210'>211</a>\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=211'>212</a>\u001b[0m     \u001b[39mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\impl.py:143\u001b[0m, in \u001b[0;36mQueuePool._do_get\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=140'>141</a>\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_inc_overflow():\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=141'>142</a>\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=142'>143</a>\u001b[0m         \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_create_connection()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=143'>144</a>\u001b[0m     \u001b[39mexcept\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/impl.py?line=144'>145</a>\u001b[0m         \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39msafe_reraise():\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:256\u001b[0m, in \u001b[0;36mPool._create_connection\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=252'>253</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_create_connection\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=253'>254</a>\u001b[0m     \u001b[39m\"\"\"Called by subclasses to create a new ConnectionRecord.\"\"\"\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=255'>256</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m _ConnectionRecord(\u001b[39mself\u001b[39;49m)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:371\u001b[0m, in \u001b[0;36m_ConnectionRecord.__init__\u001b[1;34m(self, pool, connect)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=368'>369</a>\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__pool \u001b[39m=\u001b[39m pool\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=369'>370</a>\u001b[0m \u001b[39mif\u001b[39;00m connect:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=370'>371</a>\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__connect()\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=371'>372</a>\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfinalize_callback \u001b[39m=\u001b[39m deque()\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:666\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=663'>664</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=664'>665</a>\u001b[0m     \u001b[39mwith\u001b[39;00m util\u001b[39m.\u001b[39msafe_reraise():\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=665'>666</a>\u001b[0m         pool\u001b[39m.\u001b[39mlogger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mError on connect(): \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=666'>667</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=667'>668</a>\u001b[0m     \u001b[39m# in SQLAlchemy 1.4 the first_connect event is not used by\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=668'>669</a>\u001b[0m     \u001b[39m# the engine, so this will usually not be set\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=669'>670</a>\u001b[0m     \u001b[39mif\u001b[39;00m pool\u001b[39m.\u001b[39mdispatch\u001b[39m.\u001b[39mfirst_connect:\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\langhelpers.py:70\u001b[0m, in \u001b[0;36msafe_reraise.__exit__\u001b[1;34m(self, type_, value, traceback)\u001b[0m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=67'>68</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m  \u001b[39m# remove potential circular references\u001b[39;00m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=68'>69</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mwarn_only:\n\u001b[1;32m---> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=69'>70</a>\u001b[0m         compat\u001b[39m.\u001b[39;49mraise_(\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=70'>71</a>\u001b[0m             exc_value,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=71'>72</a>\u001b[0m             with_traceback\u001b[39m=\u001b[39;49mexc_tb,\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=72'>73</a>\u001b[0m         )\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=73'>74</a>\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=74'>75</a>\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m compat\u001b[39m.\u001b[39mpy3k \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exc_info[\u001b[39m1\u001b[39m]:\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=75'>76</a>\u001b[0m         \u001b[39m# emulate Py3K's behavior of telling us when an exception\u001b[39;00m\n\u001b[0;32m     <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/langhelpers.py?line=76'>77</a>\u001b[0m         \u001b[39m# occurs in an exception handler.\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\util\\compat.py:207\u001b[0m, in \u001b[0;36mraise_\u001b[1;34m(***failed resolving arguments***)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=203'>204</a>\u001b[0m     exception\u001b[39m.\u001b[39m__cause__ \u001b[39m=\u001b[39m replace_context\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=205'>206</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=206'>207</a>\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=207'>208</a>\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=208'>209</a>\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=209'>210</a>\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=210'>211</a>\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/util/compat.py?line=211'>212</a>\u001b[0m     \u001b[39mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\pool\\base.py:661\u001b[0m, in \u001b[0;36m_ConnectionRecord.__connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=658'>659</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=659'>660</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstarttime \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime()\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=660'>661</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdbapi_connection \u001b[39m=\u001b[39m connection \u001b[39m=\u001b[39m pool\u001b[39m.\u001b[39;49m_invoke_creator(\u001b[39mself\u001b[39;49m)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=661'>662</a>\u001b[0m     pool\u001b[39m.\u001b[39mlogger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mCreated new connection \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, connection)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/pool/base.py?line=662'>663</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfresh \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\create.py:590\u001b[0m, in \u001b[0;36mcreate_engine.<locals>.connect\u001b[1;34m(connection_record)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/create.py?line=587'>588</a>\u001b[0m         \u001b[39mif\u001b[39;00m connection \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/create.py?line=588'>589</a>\u001b[0m             \u001b[39mreturn\u001b[39;00m connection\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/create.py?line=589'>590</a>\u001b[0m \u001b[39mreturn\u001b[39;00m dialect\u001b[39m.\u001b[39;49mconnect(\u001b[39m*\u001b[39;49mcargs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mcparams)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\sqlalchemy\\engine\\default.py:597\u001b[0m, in \u001b[0;36mDefaultDialect.connect\u001b[1;34m(self, *cargs, **cparams)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=594'>595</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mconnect\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39mcargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mcparams):\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=595'>596</a>\u001b[0m     \u001b[39m# inherits the docstring from interfaces.Dialect.connect\u001b[39;00m\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/sqlalchemy/engine/default.py?line=596'>597</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdbapi\u001b[39m.\u001b[39;49mconnect(\u001b[39m*\u001b[39;49mcargs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mcparams)\n",
      "File \u001b[1;32m~\\AppData\\Roaming\\Python\\Python38\\site-packages\\psycopg2\\__init__.py:122\u001b[0m, in \u001b[0;36mconnect\u001b[1;34m(dsn, connection_factory, cursor_factory, **kwargs)\u001b[0m\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=118'>119</a>\u001b[0m     kwasync[\u001b[39m'\u001b[39m\u001b[39masync_\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(\u001b[39m'\u001b[39m\u001b[39masync_\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=120'>121</a>\u001b[0m dsn \u001b[39m=\u001b[39m _ext\u001b[39m.\u001b[39mmake_dsn(dsn, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[1;32m--> <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=121'>122</a>\u001b[0m conn \u001b[39m=\u001b[39m _connect(dsn, connection_factory\u001b[39m=\u001b[39;49mconnection_factory, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwasync)\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=122'>123</a>\u001b[0m \u001b[39mif\u001b[39;00m cursor_factory \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    <a href='file:///c%3A/Users/HP/AppData/Roaming/Python/Python38/site-packages/psycopg2/__init__.py?line=123'>124</a>\u001b[0m     conn\u001b[39m.\u001b[39mcursor_factory \u001b[39m=\u001b[39m cursor_factory\n",
      "\u001b[1;31mOperationalError\u001b[0m: (psycopg2.OperationalError) connection to server at \"localhost\" (::1), port 5432 failed: FATAL:  password authentication failed for user \"root\"\n\n(Background on this error at: https://sqlalche.me/e/14/e3q8)"
     ]
    }
   ],
   "source": [
    "engine.connect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5a61a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "SELECT 1 as number;\n",
    "\"\"\"\n",
    "\n",
    "pd.read_sql(query, con=engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1c509ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "SELECT *\n",
    "FROM pg_catalog.pg_tables\n",
    "WHERE schemaname != 'pg_catalog' AND \n",
    "    schemaname != 'information_schema';\n",
    "\"\"\"\n",
    "\n",
    "pd.read_sql(query, con=engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d1f971",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('yellow_tripdata_2021-01.csv', nrows=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e76d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.tpep_pickup_datetime = pd.to_datetime(df.tpep_pickup_datetime)\n",
    "df.tpep_dropoff_datetime = pd.to_datetime(df.tpep_dropoff_datetime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f711ab67",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_sql(name='yellow_tripdata_trip', con=engine, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ca3779",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9549987c",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"\n",
    "SELECT * FROM yellow_tripdata_trip LIMIT 10\n",
    "\"\"\"\n",
    "\n",
    "pd.read_sql(query, con=engine)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcd61c63",
   "metadata": {},
   "source": [
    "```sql\n",
    "SELECT *\n",
    "FROM pg_catalog.pg_tables\n",
    "WHERE schemaname != 'pg_catalog' AND \n",
    "    schemaname != 'information_schema';\n",
    "```\n",
    "\n",
    "Source: https://www.postgresqltutorial.com/postgresql-show-tables/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffae4651",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "570feb405e2e27c949193ac68f46852414290d515b0ba6e5d90d076ed2284471"
  },
  "kernelspec": {
   "display_name": "Python 3.8.2 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
